### ISSUE TITLE
Updating from se3nn to e3nn

### ISSUE THREAD

**Issue by JoshuaMeyers** [e3nn/e3nn#160]
Hi Guys, Nice repo, and really interesting work! 

I'll admit from the start that I do not qualify as a geometry junkie. I work on drug discovery problems and we used the old version of this repo (se3nn, https://arxiv.org/abs/1807.02547) to learn protein pocket representations with great success (https://www.biorxiv.org/content/10.1101/600304v1).

I am interested in updating the network we used but I am having trouble with the translation to this new repo, I wonder if you could help point me in the right direction. In short, I have reproduced the network in e3nn by adapting the `ImageGatedConvNetwork` example, and it seems to have ~40x more parameters (`390512` (old) -> `15207552` (new)).

Any general tips/comments would be appreciated. I also have some specific questions:
- I notice that the multiplicity used by our network, is much greater than many of the examples in this repo e.g. we have a layer parameterized by Rs:`[(64,0), (64,1), (64,2), (64,3)]`. Is there something different about the higher order basis in the new version such that the number of parameters is greatly increased?
- The radial basis function has changed between the old->new versions (gaussian_window -> cosine function). Is it possible that this has many more parameters? Your tutorial suggests they should be quite similar?
- Is there still functionality to perform the convolution operation coupled to batch normalization?
- What is the relationship between steps, and stride in the Convolution operation, I am trying to use steps to downsample but it doesn't seem to have any effect on the # of params

Many thanks in advance, I am struggling through in unfamiliar territory!

Our old network:
- Model: https://github.com/BenevolentAI/DeeplyTough/blob/master/deeplytough/engine/models.py
- Model config: https://github.com/BenevolentAI/DeeplyTough/blob/master/deeplytough/scripts/train.py#L66

```
VoxelNetwork(
  (0): GatedBlock(
    (gate_act): ScalarActivation()
    (conv): SE3BNConvolution (SE3Kernel ([(8, 0)] -> [(16, 0), (16, 1), (16, 2), (16, 3), (48, 0)], size=7), eps=1e-05, momentum=0.1)
  )
  (1): GatedBlock(
    (scalar_act): ScalarActivation()
    (gate_act): ScalarActivation()
    (conv): SE3BNConvolution (SE3Kernel ([(16, 0), (16, 1), (16, 2), (16, 3)] -> [(32, 0), (32, 1), (32, 2), (32, 3), (96, 0)], size=3), eps=1e-05, momentum=0.1)
  )
  (2): GatedBlock(
    (scalar_act): ScalarActivation()
    (gate_act): ScalarActivation()
    (conv): SE3BNConvolution (SE3Kernel ([(32, 0), (32, 1), (32, 2), (32, 3)] -> [(48, 0), (48, 1), (48, 2), (48, 3), (144, 0)], size=3), eps=1e-05, momentum=0.1)
  )

...

  (5): ReLU(inplace)
  (6): BatchNorm3d(256, eps=1e-05, momentum=0.1, affine=True, track_running_stats=True)
  (7): Conv3d(256, 128, kernel_size=(1, 1, 1), stride=(1, 1, 1))
)
```

New network:
```
E3nnVoxelNetwork(
  (layers): Sequential(
    (0): Sequential(
      (0): Convolution(
        (kernel): FrozenKernel (8x1 -> 16x0,16x1,16x2,16x3,48x0)
      )
      (1): GatedBlock()
      (2): LowPassFilter()
    )
    (1): Sequential(
      (0): Convolution(
        (kernel): FrozenKernel (16x0,16x1,16x2,16x3 -> 32x0,32x1,32x2,32x3,96x0)
      )
      (1): GatedBlock()
      (2): LowPassFilter()
    )
    (2): Sequential(
      (0): Convolution(
        (kernel): FrozenKernel (32x0,32x1,32x2,32x3 -> 48x0,48x1,48x2,48x3,144x0)
      )
      (1): GatedBlock()
      (2): LowPassFilter()
    )
    (3): Sequential(
      (0): Convolution(
        (kernel): FrozenKernel (48x0,48x1,48x2,48x3 -> 64x0,64x1,64x2,64x3,192x0)
      )
      (1): GatedBlock()
      (2): LowPassFilter()
    )
    (4): Convolution(
      (kernel): FrozenKernel (64x0,64x1,64x2,64x3 -> 256x0)
    )
  )
)
```
I can share code for the adapted new network if its helpful, but it really just mirrors `ImageGatedConvNetwork` (with the ability to input custom stride, padding, and size, per layer).

**Comment by mariogeiger** [comment]
Hi,
Thanks for the interest.

> Is there something different about the higher order basis in the new version such that the number of parameters is greatly increased?
Yes, something is fundamentally different. The increase in the number of parameters is indeed due to the way the radial function is implemented. In se3nn `R(r)` was simply decomposed on a basis and that is it (gaussian of cosine does the same job). In e3nn the radial function is parameterized using a fully connected neural network. If you put no hidden layers to this network then you fall back on se3nn behavior.

> Is there still functionality to perform the convolution operation coupled to batch normalization?
I added it to my todo list. I am actually currently rewriting the entire library.

> What is the relationship between steps, and stride
steps allow to deal with non regular lattice, I know it can be useful on MRI data where measures are done in planes.

**Comment by JoshuaMeyers** [comment]
Many thanks for the informative reply, this is very helpful 🙏 . I will keep an eye on the repo and look forward to using the updated library

**Comment by JoshuaMeyers** [comment]
Quick update that removing the FCN in the radial function by setting num layers=0 reduced # params to `3844332` (i.e. by a factor of 4). I still have 10x more params than the original implementation.

**Comment by mariogeiger** [comment]
humm, indeed, on se3cnn there were also a `Jmax` what was [limiting](https://github.com/mariogeiger/se3cnn/blob/point/se3cnn/image/kernel.py#L146-L152) the spherical harmonics depending on the radius. I don't know if it explains the 10x factor

**Comment by JoshuaMeyers** [comment]
Ahh Interesting re `Jmax`.
- The newer code also takes much longer to initialise the network (~10 mins on single CPU vs <1 second for `se3nn`), could this `Jmax` be the reason for this also?

- I believe also that the new code expects the input in a different shape than the previous implementation i.e. the new code expects the channels in the last position of the input. This means I am currently doing `x.permute(0, 2, 3, 4, 1)` before the e3nn layers, and again `x.permute(0, 4, 1, 2, 3)` to return the input to its original form. Could it be that the reshaping done by `torch.einsum` in `Convolution` is different vs `se3nn` (where the se3nn version expected channels in the same position as torch `conv3d` which is preferred).

Here is my evidence for this:
Old:
```
GatedBlock(
   (gate_act): ScalarActivation()
   (conv): SE3BNConvolution (SE3Kernel ([(8, 0)] -> [(16, 0), (16, 1), (16, 2), (16, 3), (48, 0)], size=7), eps=1e-05, momentum=0.1)
)

In: torch.Size([1, 8, 24, 24, 24])
out: torch.Size([1, 256, 12, 12, 12])
```

New:
```
Sequential(
  (0): Convolution(
    (kernel): FrozenKernel (8x1 -> 16x0,16x1,16x2,16x3,48x0)
  )
  (1): GatedBlock()
)

In: torch.Size([1, 8, 24, 24, 24])
out: torch.Size([1, 4, 12, 12, 256])  # stride affects positions 1,2,3. Therefore I believe it thinks channels are in position 4.
```
By implementing the permutes above as part of `nn.Sequential`, the output `Size` is the same as se3nn at every layer.

Regarding the move towards point clouds, this makes sense, if possible I am keen to get the voxel code working since it provided really nice results in our experiments (I believe the lower number of parameters is to thank for this vs standard CNNs since standard CNNs tend to overfit for our use case – speaking of which, model is now `691840` params (roughly 2x original)).

Thanks for your help @mariogeiger

**Comment by blondegeek** [comment]
Hi @JoshuaMeyers,

I just wanted to check on the status of this issue. Are you still finding that e3nn takes substantially longer to initialize than se3cnn? I believe @mariogeiger made some changes recently that sped up several module initializations.

Best!
Tess

**Comment by JoshuaMeyers** [comment]
Hey @blondegeek, thanks for checking in on this issue. I had a look at this on the weekend. Unfortunately the large refactor means that I cannot really compare anymore. But faster sounds good! Also happy to see batchnorm make a reappearance!

In general, I was not able to setup a voxel-based network with gated blocks and batch norm analogous to that described in the original se3nn paper. This is slightly too in depth for me as a geometry rookie, part of the issue was that I am not familiar with the new string-based notation for `irreps`, and I found the `Gate` harder to use than the old preformed `GatedBlock`. I look forward to some tutorials, or example networks. I am particularly interested in describing protein structure as demonstrated in the original paper.

Happy to go with your recommendation regarding this issue, since the original purpose has drifted
